lib/libwebsockets.c

```diff
@@ -510,14 +510,28 @@ lws_close_free_wsi(struct lws *wsi, enum lws_close_status reason)
 	    !wsi->socket_is_permanently_unusable) {
 #ifdef LWS_OPENSSL_SUPPORT
 		if (lws_is_ssl(wsi) && wsi->ssl)
-               {
-                       lwsl_info("%s: shutting down SSL connection: %p (ssl %p, sock %d, state %d)\n", __func__, wsi, wsi->ssl, (int)(long)wsi->desc.sockfd, wsi->state);
+		{
+			lwsl_info("%s: shutting down SSL connection: %p (ssl %p, sock %d, state %d)\n", __func__, wsi, wsi->ssl, (int)(long)wsi->desc.sockfd, wsi->state);
 			n = SSL_shutdown(wsi->ssl);
-                       if (n == 1) /* If finished the SSL shutdown, then do socket shutdown, else need to retry SSL shutdown */
-                               n = shutdown(wsi->desc.sockfd, SHUT_WR);
-                       else
-                               lws_change_pollfd(wsi, LWS_POLLOUT, LWS_POLLIN);
-               }
+			if (n == 1) /* If finished the SSL shutdown, then do socket shutdown, else need to retry SSL shutdown */
+				n = shutdown(wsi->desc.sockfd, SHUT_WR);
+			else if (n == 0)
+				lws_change_pollfd(wsi, LWS_POLLOUT, LWS_POLLIN);
```

Comment:
@lws-team If you want to look at it, you can use [https://github.com/mmilata/libwebsockets/commits/close-connection-after-ssl-shutdown](this branch) to build a server that closes socket right after it receives ssl shutdown notification. The server spins too, though that might be caused by my changes. Also note that the server I've originally run into this problem with is based on node.js.

Here are instructions to reproduce the issue: https://gist.github.com/mmilata/0a054ef99e7235a7954a1a05c9006267